OpenCL/m25200-pure.cl

```diff
@@ -0,0 +1,218 @@
+/**
+ * Author......: See docs/credits.txt
+ * License.....: MIT
+ */
+
+//#define NEW_SIMD_CODE
+
+#ifdef KERNEL_STATIC
+#include "inc_vendor.h"
+#include "inc_types.h"
+#include "inc_platform.cl"
+#include "inc_common.cl"
+#include "inc_simd.cl"
+#include "inc_hash_sha1.cl"
+#endif
+
+#define COMPARE_S "inc_comp_single.cl"
+#define COMPARE_M "inc_comp_multi.cl"
+
+#define SNMPV3_ENGINEID_MAX 32
+#define SNMPV3_SALT_MAX 752
+
+typedef struct hmac_sha1_tmp
+{
+  u32 idx;
+  sha1_ctx_t ctx;
+
+} hmac_sha1_tmp_t;
+
+typedef struct snmpv3
+{
+  u32 salt_buf[SNMPV3_SALT_MAX];
+  u32 salt_len;
+
+  u8  engineID_buf[SNMPV3_ENGINEID_MAX];
+  u32 engineID_len;
+
+  u8 packet_number[8+1];
+
+} snmpv3_t;
+
+KERNEL_FQ void m25200_init (KERN_ATTR_TMPS_ESALT (hmac_sha1_tmp_t, snmpv3_t))
+{
+  /**
+   * modifier
+   */
+
+  const u64 gid = get_global_id (0);
+
+  if (gid >= gid_max) return;
+
+  /**
+   * base
+   */
+
+  const global u8 *pw_buf = (global u8 *) pws[gid].i;
+
+  const u32 pw_len = pws[gid].pw_len;
+
+  /**
+   * authkey
+   */
+
+  u32 idx = 0;
+
+  u32 buf[32] = { 0 };
+
+  u8 *tmp_buf = (u8 *) buf;
+
+  sha1_ctx_t ctx;
+
+  sha1_init (&ctx);
+
+  for (int i = 0; i < 64; i++)
+  {
+    tmp_buf[i] = pw_buf[idx++];
+
+    if (idx >= pw_len) idx = 0;
+  }
+
+  sha1_update_swap (&ctx, buf, 64);
+
+  tmps[gid].idx = idx;
+  tmps[gid].ctx = ctx;
+}
+
+KERNEL_FQ void m25200_loop (KERN_ATTR_TMPS_ESALT (hmac_sha1_tmp_t, snmpv3_t))
+{
+  /**
+   * base
+   */
+
+  const u64 gid = get_global_id (0);
+
+  if (gid >= gid_max) return;
+
+  const global u8 *pw_buf = (global u8 *) pws[gid].i;
+
+  const u32 pw_len = pws[gid].pw_len;
+
+  u32 idx = tmps[gid].idx;
+
+  u32 buf[32] = { 0 };
+
+  u8 *tmp_buf = (u8 *) buf;
+
+  sha1_ctx_t ctx = tmps[gid].ctx;
+
+  for (u32 j = 0; j < loop_cnt; j++)
+  {
+    for (int i = 0; i < 64; i++)
```

Comment:
This solution works, but I think we can do much better and get more speed.

All that is needed is that we allocate u32 buf[16*64]. The idea is that no matter the password length, after 64 times append of the password, the result buffer length % 64  is 0 :) 

To make this possible you need to do also: 

- fix the kernel loops to 1024 fixed in the module min and max
- for (u32 j = 0; j < loop_cnt; j++) becomes  for (u32 j = 0; j < loop_cnt; j+=64)

Additionally, we can calculate this one time in _init, then store in tmps buffer. the both variable currently in tmps get replaced with this buffer. then you can also use super fast sha1_update_64() and you do not need the swap in the loop any longer. Gutfeeling tells me this will improve performance massively.

Reply:
You also need limit password length to 64